File: Leet-222. Count Complete Tree Nodes/solution-logn.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var countNodes = function(root) {
    if (root == null) {
        return 0;
    }
    
    var leftHeight = getLeftHeight(root);
    var rightHeight = getRightHeight(root);
    
    if (leftHeight === rightHeight) {
        return Math.pow(2, leftHeight) - 1;
    }
    
    return countNodes(root.left) + countNodes(root.right) + 1;
};

var getLeftHeight = function(node) {
    var height = 0;
    while (node !== null) {
        node = node.left;
        height++;
    }
    
    return height;
};

var getRightHeight = function(node) {
    var height = 0;
    while (node !== null) {
        node = node.right;
        height++;
    }
    
    return height;
};

File: Leet-222. Count Complete Tree Nodes/solution-logn.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int countNodes(TreeNode root) {
        if (root == null) {
            return 0;
        }
        
        int leftHeight = getLeftHeight(root);
        int rightHeight = getRightHeight(root);
        
        if (leftHeight == rightHeight) {
            return (1 << leftHeight) - 1;
        } else {
            return countNodes(root.left) + countNodes(root.right) + 1;
        }
    }
    
    public int getLeftHeight(TreeNode root) {
        int height = 0;
        while (root != null) {
            height++;
            root = root.left;
        }
        
        return height;
    }
    
    public int getRightHeight(TreeNode root) {
        int height = 0;
        while (root != null) {
            height++;
            root = root.right;
        }
        
        return height;
    }
}